provisioning/baremetal_config.go

```diff
@@ -74,58 +121,121 @@ func getProvisioningNetworkMode(prov *metal3iov1alpha1.Provisioning) metal3iov1a
+	}
+	if !strings.HasSuffix(parsedURL.Path, ".qcow2.gz") && !strings.HasSuffix(parsedURL.Path, ".qcow2.xz") {
+		errs = append(errs, fmt.Errorf("the provisioningOSDownloadURL provided: %q is an OS image and must end in .qcow2.gz or .qcow2.xz", uri))
+	}
+
+	return errs
 }
 
-func validateUnmanagedConfig(prov *metal3iov1alpha1.Provisioning) error {
-	for _, toTest := range []struct {
-		Name  string
-		Value string
-	}{
-
-		{Name: "ProvisioningInterface", Value: prov.Spec.ProvisioningInterface},
-		{Name: "ProvisioningIP", Value: prov.Spec.ProvisioningIP},
-		{Name: "ProvisioningNetworkCIDR", Value: prov.Spec.ProvisioningNetworkCIDR},
-		{Name: "ProvisioningOSDownloadURL", Value: prov.Spec.ProvisioningOSDownloadURL},
-	} {
-		if toTest.Value == "" {
-			return fmt.Errorf("%s is required but is empty", toTest.Name)
-		}
+func validateProvisioningInterface(iface string) []error {
+	// Unfortunately there seems to be nothing you can really do to verify
+	// the ProvisioningInterface as it's not necessarily on this machine
+	// and could be almost anything.
+	var errs []error
+
+	if iface == "" {
+		errs = append(errs, fmt.Errorf("provisioningInterface is required but is not set"))
 	}
-	return nil
+	return errs
 }
 
-func validateDisabledConfig(prov *metal3iov1alpha1.Provisioning) error {
-	for _, toTest := range []struct {
-		Name  string
-		Value string
-	}{
-
-		{Name: "ProvisioningIP", Value: prov.Spec.ProvisioningIP},
-		{Name: "ProvisioningNetworkCIDR", Value: prov.Spec.ProvisioningNetworkCIDR},
-		{Name: "ProvisioningOSDownloadURL", Value: prov.Spec.ProvisioningOSDownloadURL},
-	} {
-		if toTest.Value == "" {
-			return fmt.Errorf("%s is required but is empty", toTest.Name)
+func validateProvisioningNetworkSettings(ip string, cidr string, dhcpRange string) []error {
+	// provisioningIP and networkCIDR are always set.  DHCP range is optional
+	// depending on mode.
+	var errs []error
+
+	// Verify provisioning ip and get it into net format for future tests.
+	provisioningIP := net.ParseIP(ip)
+	if provisioningIP == nil {
+		errs = append(errs, fmt.Errorf("The provisioningIP %q is not a valid IP", ip))
+		// I think we return here.  There are a few more tests below that require the IP be set.
+		return errs
+	}
+
+	// Verify Network CIDR
+	_, provisioningCIDR, err := net.ParseCIDR(cidr)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("%q is not a valid provisioningNetworkCIDR: %s", cidr, err))
+		// Similar thing.. need this to be valid for further tests.
+		return errs
+	}
+
+	// Ensure provisioning IP is in the network CIDR
+	if !provisioningCIDR.Contains(provisioningIP) {
+		errs = append(errs, fmt.Errorf("The provisioningIP %q is not part of the provisioningNetworkCIDR %q", ip, cidr))
+	}
+
+	// DHCP Range might not be set in which case we're done here.
+	if dhcpRange == "" {
+		return errs
+	}
+
+	// We want to allow a space after the ',' if the user likes it.
+	dhcpRange = strings.ReplaceAll(dhcpRange, " ", "")
```

Comment:
This will remove spaces that appear in other places in the string, so that a value like `192.168  .20.1,192.168. 20.100` is treated as valid. Should it be?

Reply:
True.  Fixed.